OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnCombatantOwner.cs

```diff
@@ -0,0 +1,52 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Grants a condition if owner is combatant.")]
+	public class GrantConditionOnCombatantOwnerInfo : TraitInfo
+	{
+		[FieldLoader.Require]
+		[GrantedConditionReference]
+		[Desc("The condition to grant.")]
+		public readonly string Condition = null;
+
+		public override object Create(ActorInitializer init) { return new GrantConditionOnCombatantOwner(init.Self, this); }
+	}
+
+	public class GrantConditionOnCombatantOwner : INotifyCreated, INotifyOwnerChanged
+	{
+		readonly GrantConditionOnCombatantOwnerInfo info;
+
+		int conditionToken = Actor.InvalidConditionToken;
+
+		public GrantConditionOnCombatantOwner(Actor self, GrantConditionOnCombatantOwnerInfo info)
+		{
+			this.info = info;
+		}
+
+		void INotifyCreated.Created(Actor self)
+		{
+			if (!self.Owner.NonCombatant)
+				conditionToken = self.GrantCondition(info.Condition);
+		}
+
+		void INotifyOwnerChanged.OnOwnerChanged(Actor self, Player oldOwner, Player newOwner)
+		{
+			if (!newOwner.NonCombatant && conditionToken == Actor.InvalidConditionToken)
+				conditionToken = self.GrantCondition(info.Condition);
+			else if (newOwner.NonCombatant && conditionToken != Actor.InvalidConditionToken)
```

Comment:
This should probably revoke the old condition even if the new owner isn't a combatant.